### PR TITLE
Delay unmounting elements until the end of the frame

### PR DESCRIPTION
**Pull request by abarth**
This patch prepares us to move these elements around in the tree.

### CODE REVIEW

File: sky/packages/sky/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -413,6 +415,38 @@ enum _ElementLifecycle {
   defunct,
 }
 
+class _FreeElements {
+  bool _locked = false;
+  final Set<Element> _elements = new Set<Element>();
+
+  void _finalize(Element element) {
+    element.unmount();
```
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/framework.dart]: is there a way to distinguish a node that should be in this list from one that should not? e.g. does it have its parent removed or something? If so, assert it here. If not, maybe that should be a new lifecycle state?

File: sky/packages/sky/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -468,15 +502,6 @@ abstract class Element<T extends Widget> implements BuildContext {
   /// Calls the argument for each child. Must be overridden by subclasses that support having children.
   void visitChildren(ElementVisitor visitor) { }
 
-  /// Calls the argument for each descendant, depth-first pre-order.
```
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/framework.dart]: make sure none of the demos or tests use this... I'm pretty sure I've tried to use it in the past (though I may have fallen back on visitChildren).

Comment (abarth) [inline sky/packages/sky/lib/src/widgets/framework.dart]: There aren't any callers in the repo anymore.

File: sky/packages/sky/lib/src/widgets/framework.dart [hunk 3]
```diff
@@ -1170,6 +1194,7 @@ abstract class RenderObjectElement<T extends RenderObjectWidget> extends Element
       _ancestorRenderObjectElement.removeChildRenderObject(renderObject);
       _ancestorRenderObjectElement = null;
     }
+    _slot = null;
```
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/framework.dart]: be consistent, in this and the superclass implementation of this method, about whether _slot is nulled out first or last.

File: sky/packages/sky/lib/src/widgets/framework.dart [hunk 4]
```diff
@@ -413,6 +415,38 @@ enum _ElementLifecycle {
...
+          _finalize(element);
+      } finally {
+        _elements.clear();
+        _locked = false;
+      }
+    });
+  }
+
+  void add(Element element) {
+    assert(!_locked);
```
Comment (Hixie) [inline sky/packages/sky/lib/src/widgets/framework.dart]: change the lifecycle state of the element, and mark it not _dirty, when adding

**Comment by Hixie** [comment]
LGTM

**Pull request merged**